Fix error handling in reviewvote, only show buttons when logged in @salty-horse

salty-horse opened the pull request:
Fixes #1192.

* jsonSend didn't handle some errors correctly. It didn't show the alert message when there was an HTTP error, and it always called the callback. IIRC the intent is to only call the callback when there are no HTTP errors. (The callback can't tell if it was successful or not). Now it will still call it if there's an `error` field in the JSON body.
* When a user is logged out, they will not be presented with the review vote controls.

dfabulich reviewed (changes requested):
In #1192, you wrote:

> The voting UI should be hidden when you're not logged in. (IMO - I prefer that to trying to do the action and being redirected to a login page.)

I think I don't agree with that approach. When logged out, there should be _some_ invitation to login to use features, rather than hiding the features completely.

For example, "Have you played this game?" says, "You can rate this game, record that you've played it, or put it on your wish list after you log in." We can put a message like that here, e.g. "You can log in to rate this review, mute this user, or add a comment."

That doesn't always fit in the UI, and so a good enough second-choice option is to leave the links in place but require login, as we do for "Add a Comment".

Regardless, this PR is currently doing something kinda weird, hiding _just_ the review vote options, but showing "More Options" (all of which require login, except for "Direct Link") and Add a Comment (which redirects to login).

(And, why would it be called "More Options" when the review vote options are hidden…?)

salty-horse commented:
Mostly agreed.

I don't think every review should have a "call to action" to login under it, when there can be several reviews on a page.

I missed that detail about "More Options" only having login-requiring actions. I originally put the "Direct Link" in the "More" menu and I regret it. Got any suggestions on where to place it? Usually permalinks are placed on the timestamps but that's not ideal.

dfabulich commented:
> I don't think every review should have a "call to action" to login under it, when there can be several reviews on a page.

I think it's OK (good) to show the call-to-action message next to every review, (as long as the message itself is brief) because I think it's unlikely that we'd show the call to action multiple times on the same screen. The median review is ~600-700 words, a few dozen lines long. So I think the user will likely see it only once (or at most twice) in any given screenful.

If we're hiding "More Options" when logged out, then I think the thing to do is to add the permalink to the timestamp, but I think it's OK to leave it in "More Options", too. It's redundant, but it's not occupying extra screen space, and the redundancy would make the permalink a little bit easier to discover.

salty-horse commented:
Rebased over main to include the plonk->mute changes.

1. Added a "Log in to do stuff" message with your suggested phrasing. The redirect after login leads to the specific review, rather than the game page.
2. The "more options" box only appears when you're logged in.
3. The timestamps in reviews is now a permalink with the "silent" style, so it's not blue and only has an underscore on hover. That permalink is only on reviews with text - should they be included in rating-only reviews? The permalink is always there, even when viewing a specific review, because I expect permalinks to always be there in case I want to copy them.

dfabulich commented:
> That permalink is only on reviews with text - should they be included in rating-only reviews?

I guess it would be nice, but I wouldn't hold up this PR over it.
